core/trino-main/src/test/java/io/trino/sql/planner/assertions/JoinMatchPatternBuilder.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.assertions;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.sql.planner.iterative.rule.test.PlanBuilder;
+import io.trino.sql.planner.plan.JoinNode;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.node;
+import static io.trino.sql.planner.plan.JoinNode.DistributionType;
+import static io.trino.sql.planner.plan.JoinNode.EquiJoinClause;
+import static io.trino.sql.planner.plan.JoinNode.Type;
+import static io.trino.sql.tree.ComparisonExpression.Operator.EQUAL;
+import static java.util.Objects.requireNonNull;
+
+public class JoinMatchPatternBuilder
+{
+    private Type joinType;
+    private List<ExpectedValueProvider<EquiJoinClause>> equiCriteria;
```

Comment:
make it `Optional<ExpectedValueProvider<EquiJoinClause>> equiCriteria = Optional.empty()` (here and other fields)

We prefer to use explicit `Optionals` rather than nulls apart from perf critical places.
If some field can actually be null, mark it as `@Nullable`